test/inductor/test_online_softmax.py

```diff
@@ -0,0 +1,195 @@
+                .matmul(v)
+            )
+
+        opt_f = torch.compile(f)
+        ref = f(q, k, v)
+        act, (code,) = run_and_get_code(opt_f, q, k, v)
+        self.assertTrue(torch.allclose(ref, act, atol=1e-2, rtol=1e-2))
+        self.assertTrue("aten._scaled_dot_product_" in code)
+
+    @parametrize("nrow", [2, 2048])
+    @parametrize("dim", [-1, 0, 1])
+    def test_prepare_softmax(self, dim, nrow):
+        x = torch.randn(nrow, 2048, dtype=torch.bfloat16, device=GPU_TYPE)
+        act, (code,) = run_and_get_code(torch.compile(_prepare_softmax), x, dim)
+        ref = _prepare_softmax(x, dim)
+        self.assertTrue(same(ref, act, tol=1e-2))
+
+        if nrow == 2048 and dim == 0:
+            # split reduction is triggered. We have multiple kernels
+            self.assertTrue(code.count("def triton") >= 2)
+        else:
+            if nrow == 2 and dim == 0:
+                # persistent reduction triggered
+                expected_num_loop = 0
+            else:
+                # A single loop due to online softmax
+                expected_num_loop = 1
+            self.assertEqual(code.count("for r0_offset in"), expected_num_loop)
+
+    def test_split_reduction(self):
+        """
+        We don't split online_softmax_reduce for now. Check
+        'Split online_softmax_reduce' note in the code.
+
+        When a split is promsing, we fallback for now.
+
+        This is just a manual example rather than something we
+        see in practice.
+        """
+        # tensor shape to trigger split reduction
+        x = torch.randn(1, 2**20, dtype=torch.bfloat16, device=GPU_TYPE)
+        ref = torch.softmax(x, dim=-1)
+        act, (code,) = run_and_get_code(torch.compile(torch.softmax), x, dim=-1)
+        self.assertTrue(torch.allclose(ref, act, atol=1e-3, rtol=1e-3))
+        self.assertTrue(code.count("def triton") >= 2)
+        self.assertTrue("online_softmax_reduce" not in code)
+
+    @parametrize("dtype", [torch.bfloat16, torch.half, torch.float32])
+    def test_acc_with_fp64(self, dtype):
+        x = torch.randn(32768, 50257, device=GPU_TYPE, dtype=dtype)
+
+        ref_fp64 = _prepare_softmax(x.to(dtype=torch.float64), dim=-1)
+        ref = _prepare_softmax(x, dim=-1)
+        res, (code,) = run_and_get_code(torch.compile(_prepare_softmax), x, dim=-1)
+        self.assertTrue("online_softmax_reduce" in code)
+
+        # Max should be exactly equal
+        self.assertEqual(ref[0], res[0])
+        self.assertEqual(ref[0].to(dtype=torch.float64), ref_fp64[0])
+
+        ref_error = rmse(ref_fp64[1], ref[1]).item()
+        res_error = rmse(ref_fp64[1], res[1]).item()
+
+        # My local tests even shows a smaller res_error:
+        #   ref_error=2.1065, res_error=2.1028
+        # for bf16
+        #   ref_error=0.2611, res_error=0.2609
+        # for fp16
+        #   ref_error=0.0001, res_error=0.0001
+        # for fp32
+        print(f"{ref_error=:.4f}, {res_error=:.4f}")
```

Comment:
@Chillee the numerics of online softmax is even slightly better in my tests compared to a fp64 baseline